src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,133 @@
+	///
+	/// @returns A pointer to the allocated struct.
+	MumbleChannel_t* (*allocateMumbleChannel)();
+
+	/// Frees a MumbleChannel that has been allocated via allocateMumbleChannel.
+	///
+	/// @param channel A pointer to the channel struct that should get freed
+	void (*freeMumbleChannel)(MumbleChannel_t *channel);
+
+	
+
+	
+	// -------- Getter functions --------
+
+	/// Fills in the connection ID of the server the user is currently active on (the user's audio output is directed at).
+	///
+	/// @param[out] connection A pointer to the memory location the ID should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then it is valid to access the
+	/// 	value of the provided pointer
+	uint32_t getActiveServerConnection(MumbleConnection_t *connection);
+
+	/// Fills in the information about the local user.
+	///
+	/// @param connection A reference to which server-connection should be considered.
+	/// @param[out] user The user struct allocated via allocateMumbleUser whose fields will be set by this function.
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the fields of the
+	/// 	passed struct may be accessed.
+	uint32_t (*getLocalUser)(MumbleConnection_t connection, MumbleUser_t *user);
+
+	/// Fills in the information about the user with the provided ID, if such a user exists.
+	///
+	/// @param userID The respective user's ID
+	/// @param connection A reference to which server-connection should be considered.
+	/// @param[out] user The user struct allocated via allocateMumbleUser whose fields will be set by this function.
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the fields of the
+	/// 	passed struct may be accessed.
+	uint32_t (*getUser)(MumbleConnection_t connection, uint32_t userID, MumbleUser_t *user);
+
+	///	Fills in the information about the channel with the given ID, if such a channel exists.
+	///
+	/// @param connection A reference to which server-connection should be considered.
+	/// @param channelID The respective channel's ID
+	/// @param[out] channel The channel struct allocated via allocateMumbleChannel whose fields will be set by this function.
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the fields of the passed
+	/// 	struct may be accessed.
+	uint32_t (*getChannel)(MumbleConnection_t connection, int32_t channelID, MumbleChannel_t *channel);
+
+	/// Fills in the amount of clients that are currently connected to the server.
+	///
+	/// @param connection A reference to which server-connection should be considered.
+	/// @param[out] A pointer to the memory the amount shall be written
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then it is valid to access the
+	/// 	value of the provided pointer.
+	uint32_t (*getAmountOfConnectedClients)(MumbleConnection_t connection, size_t *amount);
```

Comment:
I think "Number" might be better than Amount.
Why do we need this function? We will need a way to retrieve every client on server/in channel anyway, such a API function would already give us the number.

Reply:
Because I was thinking about having the plugin allocate an array for the client list. And in order to know how much memory the plugin needs to allocate for this, I introduced this function

Reply:
Would be much easier to let mumble allocate the memory, and just provide a generic `free(void*)` function via API to let the plugin free the memory.

Reply:
But how can I free a void pointer without a constructor? Consider the `MUmbleClient` struct. It stores a `const char *` which eventually points to a String that had to be allocated. If freeing the struct, only the pointer gets freed. The allocated string doesn't.

And without knowing the type of the pointer that should be freed, how could I catch these cases in the function?

Reply:
Could store pointer->deletefunction mapping in a map internally. Then on free look up what destruction code to call in the map, cast the void* over to it and do cleanup.